enhancements/machine-config/mco-flattened-ignition.md

```diff
@@ -0,0 +1,191 @@
+proposals around a future [declarative network configuration](https://github.com/openshift/enhancements/pull/399)
+although we need to ensure the approach taken in each case is aligned.
+
+### User Stories
+
+#### Story 1
+
+As a UPI user, I would like to perform network configuration in the same
+way as other customization, e.g via MachineConfig manifests or ignition.
+
+#### Story 2
+
+As a baremetal IPI user, I need to deploy in an environment where network
+configuration is required before access to the controlplane network is possible.
+
+Specifically I wish to deploy in an environment where a bonded pair of nics
+are used for the controlplane, on a non-default VLAN (a common configuration
+for baremetal where resilience is required but you want to minimise the total
+number of NICs needed).
+
+## Design Details
+
+The MCO currently maintains two ignition artifacts, a managed instance of the
+pointer ignition config, stored in a Secret, and the rendered config stored
+in a MachineConfig object (both per role, not per machine).
+
+Only the rendered config is accessible via the MCS (although not currently from within
+the cluster ref https://github.com/openshift/machine-config-operator/issues/1690)
+so they may be consumed during deployment.
+
+The proposal is to create a third managed artifact, which is the flattened
+combination of the pointer and rendered ignition configs, so that when
+required it may be consumed instead of requiring runtime download of the rendered
+config via the pointer URL.
+
+This additional artifact will be an additional MachineConfig resource, that is
+regenerated every time the existing rendered config is updated by the machine
+config render controller.  It will contain a combination of the pointer ignition
+and the rendered config.
+
+This new resource will be made available via a new API path in the MCS e.g
+`config/master/flattened` (or alternatively a query string e.g `config/master?flattened=true`)
+
+To work around the issue that the MCS is not accessible from the the cluster for
+the baremetal machine-api actuator, that component will retrieve the rendered
+config directly via the k8s API instead (TODO verify the RBAC for the machine
+controller will permit this due to different namespaces).
+
+### Test Plan
+
+  * Ensure no regressions in existing deployments via existing e2e-metal-ipi CI coverage
+  * Prove we can do deployments where the controlplane network requires configuration, such as that described in [previous bug reports](https://bugzilla.redhat.com/show_bug.cgi?id=1824331) - we can add a periodic CI job based on e2e-metal-ipi to test this.
+
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+### Upgrade / Downgrade Strategy
+
+This change only impacts the day1 deployment, so that deployment becomes
+possible without workarounds, so there should be no upgrade impact.
+
+### Version Skew Strategy
+
+## Alternatives
```

Comment:
Another alternative could be to just merge the installer generated ignition config into the rendered config (I guess we'd need to create a MachineConfig object derived from the installer generated user-data secret, then rely on the existing MergeMachineConfigs logic in MCO.

@runcom that would also solve the issue with the management of the pointer ignition, since we'd convert any data from the installer-created ignition to a MachineConfig, thus the pointer config can just be statically tempated as in your recent implementation?

Reply:
To clarify this idea, something like:

- Installer creates master-user-data (as it does today, [may contain user ignition customizations](https://github.com/openshift/installer/blob/master/docs/user/customization.md#os-customization-unvalidated))
- MCO parses master-user-data and creates a MachineConfig with the ignition data
- MCO generates master-user-data-managed from template (as implemented in https://github.com/openshift/machine-config-operator/pull/1792 but reverted via https://github.com/openshift/machine-config-operator/pull/2126 due to breaking the installer interface)
- Any installer provided customizations show up in the existing rendered config (same as if the user had provided them via MachineConfig manifests)

We could potentially flag some warning when this happens, and use it as a migration path to eventually deprecate the direct ignition-configs customization in favor of MachineConfig manifests?

This would also solve the issue for IPI baremetal without any MCO API changes, since we could consume the existing rendered config directly (we did this previously ref https://github.com/openshift/installer/pull/3276 but that was also reverted for the same reason as above)

@cgwalters @crawford any thoughts?

Reply:
I guess there are some RBAC considerations:

- The current user-data secret ends up in the openshift-machine-api namespace - https://github.com/openshift/machine-config-operator/pull/1792 worked around that by generating the pointer ignition manifest in the right namespace
- We could adjust the installer to write future user-data to the openshift-machine-config-operator namespace (so it can be read and converted into a  MachineConfig resource)
- On upgrade the plan from https://github.com/openshift/machine-config-operator/pull/1792 remains, e.g just reference the existing non-managed user-data secret?

Reply:
I looked at the code, and it's probably simplest to have the installer generate a MachineConfig manifest that contains any config provided via `create ignition-configs`

That avoids any changes to the MCO (other than restoring the reverted pointer-ignition change from @runcom) and potentially allows us to warn the user when config is provided via that interface.

Reply:
I think I'm generally in favour of the MCO managing the stub configs coming from installer in one way or another, and this approach will probably work. I think we should keep the following points in mind:
1. Since the stub configs were never changed before we never considered any dependency between the MAO and the MCO. This method would likely (as I understand it) introduce an ordering where the MCO has to process the stub config first before machines can be booted.
2. We implicitly supported "per-node" configuration with the stub config, for example after you generated the stub config, you could configure each node's networking separately before boot with a different static IP via the ignition stub, and the MCO would be fine with it since it had no understanding of what existed in the stub. I guess for this example, the generated configs with `openshift-installer create ignition-configs` would have been supplied to the nodes manually. Is this a case we care about?

Reply:
good points, I care about solving point number 2 as it seems we really never wanted to advertise the use of that (last we checked with @crawford )

Reply:
> I think I'm generally in favour of the MCO managing the stub configs coming from installer in one way or another, and this approach will probably work. I think we should keep the following points in mind:
> 
>     1. Since the stub configs were never changed before we never considered any dependency between the MAO and the MCO. This method would likely (as I understand it) introduce an ordering where the MCO has to process the stub config first before machines can be booted.

I think we can avoid this if instead the installer creates a MachineConfig manifest with any user customizations it finds on the `create cluster` phase, and we could either skip generating this or leave it empty in the case where the stub config output via the `create ignition-configs` phase is unmodified?

We can also warn the user if the decision is made to deprecate the modification of the stub config with this approach (which may be harder if we have the MCO process the stub config, in addition to the ordering concern you raise above).
 
>     2. We implicitly supported "per-node" configuration with the stub config, for example after you generated the stub config, you could configure each node's networking separately before boot with a different static IP via the ignition stub, and the MCO would be fine with it since it had no understanding of what existed in the stub. I guess for this example, the generated configs with `openshift-installer create ignition-configs` would have been supplied to the nodes manually. Is this a case we care about?

Good point - IIUC this has only ever worked for the UPI case, where you download the stub config from the installer `create ignition-configs` phase, then host some per-host modified configs somewhere outside of the cluster?

In the IPI case, there's only a single secret per role, so per-node config is not currently supported via that workflow.

So, I think to ensure both workflows continue to work the same as today, we just need to ensure that the new MachineConfig object either isn't generated or is empty at the `create ignition-configs` stage, and that we can detect and re-generate that asset in the case where some user-customization happened between `create ignition-configs` and `create cluster` (I guess we can compare the assets loaded from file with that generated inside the installer).